### DISCUSSION TITLE
When should I use plugins, composables and utils?

### DISCUSSION THREAD

**Discussion by RomanSkrypnik** [nuxt/nuxt#23694] [Questions]
Frankly speaking, I'm a little bit confused about using plugins, composables and utils in nuxt

If usage of plugins can be explained by app initialization, directives creation and third party plugins usage then I don't see essential difference between composables and utils

Documentation says:
![image](https://github.com/nuxt/nuxt/assets/44501848/a3edea80-fde5-4ee6-ad31-8b5a7139b8d8)
What kind of semantic distinction is exactly mentioned?
Could you please describe?

**Comment by jarmos-san** [comment]
Stumbled across this question while looking for an answer myself! Here's what I understand based on limited details provided on the subject elsewhere, a little bit of ChatGPT and of course personal experience.

1. Use the `utils` directory for any **"pure"** JavaScript/TypeScript functions which can exist and run in a Node.js environment without depending on Nuxt.js or Vue.js.
2. Use the `composables` directory for "**stateful**" functions which require Vue/Nuxt's reactivity capabilities.
3. Use the `plugins` directory if your function needs to hook into the Nuxt.js app lifecycle (not just the initialization phase).

**Comment by initial-d** [comment]
The difference is mostly about coupling, not about what JavaScript is technically able to do.

I usually split them like this:

**utils**: plain reusable functions

Use `utils/` for code that does not need Vue or Nuxt runtime context.

Examples:

```ts
export function formatPrice(value: number) { ... }
export function slugify(input: string) { ... }
export function groupBy<T>(items: T[], key: keyof T) { ... }
```

A good test is: could this function live in a normal TypeScript package and run in Node, a test runner, or a script without a Nuxt app? If yes, it is probably a util.

**composables**: reusable Vue/Nuxt reactive logic

Use `composables/` when the function depends on Vue reactivity or Nuxt composables/context.

Examples:

```ts
export function useCart() {
  const items = useState('cart:items', () => [])
  const total = computed(() => ...)
  return { items, total }
}

export function useCurrentUser() {
  return useAsyncData('current-user', () => $fetch('/api/me'))
}
```

This code is not just a helper function; it participates in Vue/Nuxt's reactive lifecycle. It may call `ref`, `computed`, `watch`, `useState`, `useRoute`, `useRuntimeConfig`, `useAsyncData`, etc.

**plugins**: app setup and injection

Use `plugins/` when something must be installed into the Nuxt app once, before components/composables use it.

Examples:

```ts
export default defineNuxtPlugin((nuxtApp) => {
  const api = createApiClient(...)
  return {
    provide: { api },
  }
})
```

Good plugin use cases:

- registering a Vue plugin or directive;
- setting up a third-party SDK once;
- providing `$api`, `$analytics`, `$i18n`, etc.;
- hooking into Nuxt app lifecycle.

A small practical decision tree:

- No Vue/Nuxt dependency? Put it in `utils/`.
- Uses reactivity or Nuxt composables and is called from components/pages/other composables? Put it in `composables/`.
- Needs to run once during app initialization or inject something into the app? Put it in `plugins/`.

The semantic distinction is useful because it keeps dependencies flowing in one direction. A util should not secretly depend on app context. A composable can depend on utils. A plugin can create app-level services that composables/components later consume. That makes the code easier to test and avoids accidentally calling Nuxt-only APIs in places where no Nuxt app exists yet.
